api/turing/batch/controller/ensembling.go

```diff
@@ -0,0 +1,268 @@
+package batchcontroller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	apisparkv1beta2 "github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta2"
+	"github.com/gojek/turing/api/turing/cluster"
+	"github.com/gojek/turing/api/turing/config"
+	"github.com/gojek/turing/api/turing/log"
+	"github.com/gojek/turing/api/turing/models"
+	"github.com/gojek/turing/api/turing/service"
+	apicorev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+)
+
+// SparkApplicationState is the state of the spark application
+type SparkApplicationState string
+
+const (
+	ensemblingJobApplicationPath = "local:///home/spark/batch-ensembler/main.py"
+	// SparkApplicationStateRunning is when the spark application is still running
+	SparkApplicationStateRunning = SparkApplicationState(iota)
+	// SparkApplicationStateCompleted is when the spark application has completed its run
+	SparkApplicationStateCompleted
+	// SparkApplicationStateFailed is when the spark application has failed its run
+	SparkApplicationStateFailed
+	// SparkApplicationStateUnknown is when the spark application state is unknown
+	SparkApplicationStateUnknown
+)
+
+// CreateEnsemblingJobRequest is a request to run an ensembling job on Kubernetes.
+type CreateEnsemblingJobRequest struct {
+	EnsemblingJob *models.EnsemblingJob
+	Labels        map[string]string
+	ImageRef      string
+	Namespace     string
+}
+
+// EnsemblingController is an interface that exposes the batch ensembling kubernetes controller.
+type EnsemblingController interface {
+	Create(request *CreateEnsemblingJobRequest) error
+	GetStatus(namespace string, ensemblingJob *models.EnsemblingJob) (SparkApplicationState, error)
+}
+
+type ensemblingController struct {
+	clusterController cluster.Controller
+	mlpService        service.MLPService
+	sparkInfraConfig  *config.SparkInfraConfig
+}
+
+// NewBatchEnsemblingController creates a new batch ensembling controller
+func NewBatchEnsemblingController(
+	clusterController cluster.Controller,
+	mlpService service.MLPService,
+	sparkInfraConfig *config.SparkInfraConfig,
+) EnsemblingController {
+	return &ensemblingController{
+		clusterController: clusterController,
+		mlpService:        mlpService,
+		sparkInfraConfig:  sparkInfraConfig,
+	}
+}
+
+func (c *ensemblingController) GetStatus(
+	namespace string,
+	ensemblingJob *models.EnsemblingJob,
+) (SparkApplicationState, error) {
+	sa, err := c.clusterController.GetSparkApplication(namespace, ensemblingJob.Name)
+	if err != nil {
+		return SparkApplicationStateUnknown, fmt.Errorf("failed to retrieve spark application %v", err)
+	}
+
+	state := sa.Status.AppState.State
+	if state == apisparkv1beta2.CompletedState {
+		return SparkApplicationStateCompleted, nil
+	}
+
+	if state == apisparkv1beta2.FailedState {
+		return SparkApplicationStateFailed, nil
+	}
+
+	if state == apisparkv1beta2.UnknownState {
+		return SparkApplicationStateUnknown, nil
+	}
+
+	return SparkApplicationStateRunning, nil
```

Comment:
Yeah, I don't think we need show the specifics on what it is actually doing on Kubernetes since this is an abstraction layer. To the user it's either running or not running/failed/completed. On Spark's side it will keep retrying until the retry count has been reached which means we will show to the user that the ensembling job has failed.

Reply:
My opinion is that the interim states are not as important as the terminal states.

Reply:
Ok, that makes sense.